### ISSUE TITLE
Add blocking events to the Admin interface to allow people to write (external) plugins.

### ISSUE THREAD

**Issue by 1Hyena** [bitcoinclassic/bitcoinclassic#250]
This is more of an idea for a new feature rather than an issue with any existing software.

I think Bitcoin needs a reference client that is politically neutral by default and allows the full node operator to easily make any kind of a soft fork for their personal use. Typically in software development context such ability could be referred to as a plugin system. I don't think everyone should make their own fork of the original Satoshi's Bitcoin Wallet just to change a couple of small rules.

Example use case would be to add custom filters between the raw incoming data from the network and the mempool of TXs. The power of the full node is to decide which TXs and blocks they wish to relay and which not. So let's bring that power to the full node operators. I would be more than happy to write a custom block and TX filter for myself that would reject SegWit TXs and blocks containing or allowing to contain SegWit TXs. This is such a small adjustment that it does not seem rational to make my own branch of the reference client just for that.

With such a plugin system everyone could make simplex soft forks rather easily. It is one more reason for people to run full nodes and it incentivizes full node operators to maintain a well-connected node on the network.

**Comment by zander** [comment]
This is not a bad idea.

I personally think that the full node software should do the minimum amount of things, and do those really really well. And this makes a plugin system go out of scope very quickly.

I do agree with your suggestions that this would be great to have and as such I was thinking about how this could be done cleanly.

An idea came up where the full node uses Classics [AdminServer](https://bitcoinclassic.com/devel/AdminServer.html) to notify listeners of new transactions that are accepted in the mempool.

It would be much easier to have such a "plugin" interface based on top of a simple protocol where any application in any programming language can then register itself and "filter" transactions just like you suggest.

I'd welcome such innovation.

**Comment by 1Hyena** [comment]
Yes, perhaps there should be some kind of TX and block propagation proxy/wrapper as a separate application which will essentially sandbox the full node. The full node itself would have to be configured to connect ONLY to that proxy which will most likely run on localhost. The proxy would have to do the p2p networking. I think most full node implementations allow the node to connect to only a predefined set of hosts (-connect command line parameter?) This approach should comply to the principles of modular design quite well.

**Comment by zander** [comment]
that is roughly the same idea I wrote about, but the suggestion I wrote is much simpler as it doesn't require you to proxy p2p traffic (which is not exactly trivial as the p2p layer is very proprietary).

**Comment by 1Hyena** [comment]
Perhaps a less portable way of achieving this would be to capture the packets of bitcoind with the help of the OS similarly to how wireshark does it. The custom soft fork filters could be applied on that level then. I'm not familiar with AdminServer though. But anyway, wouldn't it make sense to start a Bitcoin hub project which main purpose is to relay/proxy the p2p data without doing any validation and without storing anything on disk? Such a hub would be a perfect application where to stick the soft fork rules.

edit:
This project would essentially be(come) a Bitcoin Firewall :D

**Comment by zander** [comment]
> I'm not familiar with AdminServer though.

There is some more information about it here; https://bitcoinclassic.com/devel/AdminServer.html

**Comment by zander** [comment]
There are more people that are interested in this topic, so I want to suggest an approach that I think will be useful.

Using a networked client ("The Plugin") that communicates with Bitcoin Classic over the Admin Server (unix socket) that client can subscribe to events. Events like a block that's coming in.

Then when Classic gets a block it validates its header and then notifies the plugin via the admin server. Probably inserting some meta data like the header, the amount of transactions and the size of the block. The admin server replies either Ok or Cancel. Causing this block to be accepted or rejected.

Notice that The Plugins lives in a different process, or even on a different machine, from the Bitcoin Classic node. I think that is a smart design to keep the client lean and easy to maintain.

**Comment by 1Hyena** [comment]
@zander 
this is an excellent idea. If the plugin could potentially even live on a different machine it could easily be shared as a "service" in a company's private network. That way a single machine will dictate the soft fork policies for the whole company that could be running many full nodes. could turn useful in some cases.

perhaps this architecture would also make it possible to easily implement your own mempool rules to determine which unconfirmed TX chains are kept in mempool and which ones get kicked out when no more memory could be allocated for the mempool. Miners could then implement their own custom logic more easily. For example, a miner would be interested in confirming the chain on TXs that pays the most fees per byte even if the oldest TX in that chain is dust.

What is more, making the plugin a completely separate process allows easier contribution to the project because the contributors don't have to dive into the reference client's own source code.
